### ISSUE TITLE
Add simple external state management to ReactSpecimen

### ISSUE THREAD

**Issue by mhintz** [interactivethings/catalog#88]
Per our discussion earlier this morning.

Many React components expect to be embedded in a context which handles their state for them. In particular, it's great to create stateless React components which simply call callbacks on user interaction, and expect to have something like a `value` prop updated whenever they should change their rendered state.

However, when documenting this type of component in Catalog, such components are often not interactive, because there's nothing updating their `value`s.

This feature request is for ReactSpecimen to include simple state management, so that a documented component can access properties on `this.state` and can call `this.setState` inside a callback.

For example:

```
<Input value={this.state.value} onChange={(evt) => this.setState({ value: evt.target.value })} />
```

**Comment by naehrstoff** [comment]
When documenting this, it's important to note that this is for the Markdown renderer – it was unclear to me at first why this is needed.

I'm not sure whether `this` is the right variable name to use. In the implementation it is `this` of the parent `ReactSpecimen` container, but this is not visible in the Markdown code. Furthermore, what is `this`: is it the real parent's `this` with all additional properties and methods or is it a dumbed down `this` that only exposes `this.state` and `this.setState`? Due to this unclarity and to prevent undocumented backdoors into the parent class I would suggest to either use the name “context” with `context.getState()` and `context.setState()` or just provide two functions `getState()` and `setState()`. I prefer the former.

**Comment by herrstucki** [comment]
Added in decde86. Related to #87.

The tricky part was how to specify initial state. I think I found a nice solution where we separate specimen props from content with `---` (similar to yaml frontmatter in md documents):


    ```react
    initialState:
      foo: 'bar'
    ---
    <button onClick={() => setState({foo: 'baz'})}>{state.foo}</button>
    ```

This works for all specimens, but is most useful react, html, hint, and code specimens, where the specimen body (i.e. what's inside the code block) is used as the specimen output.

Possible improvement: just use `state` instead of `initialState` (like in a component constructor you use `this.state = {...}`).
